src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java

```diff
@@ -0,0 +1,235 @@
+                            statusCode, request, requestStartMs, responseHeaders));
+        }
+
+        byte[] responseContents;
+        if (httpResponse.getContentLength() == -1) {
+            // Add 0 byte response as a way of honestly representing a
+            // no-content request.
+            responseContents = new byte[0];
+        } else {
+            responseContents = httpResponse.getContentBytes();
+        }
+
+        if (responseContents != null) {
+            onResponseRead(
+                    requestStartMs,
+                    statusCode,
+                    httpResponse,
+                    request,
+                    callback,
+                    responseHeaders,
+                    responseContents);
+            return;
+        }
+
+        final InputStream inputStream = httpResponse.getContent();
+        Runnable run =
+                new Runnable() {
+                    @Override
+                    public void run() {
+                        byte[] finalResponseContents = new byte[0];
+                        try {
+                            finalResponseContents =
+                                    NetworkUtility.inputStreamToBytes(
+                                            inputStream, httpResponse.getContentLength(), mPool);
+                        } catch (IOException e) {
+                            onRequestFailed(
+                                    request, callback, e, requestStartMs, httpResponse, null);
+                        } catch (ServerError serverError) {
+                            // This should never happen since we already check if inputStream is
+                            // null
+                            throw new RuntimeException(serverError);
+                        }
+                        onResponseRead(
+                                requestStartMs,
+                                statusCode,
+                                httpResponse,
+                                request,
+                                callback,
+                                responseHeaders,
+                                finalResponseContents);
+                    }
+                };
+        mBlockingExecutor.execute(run);
+    }
+
+    /* Method to be called after a failed network request */
+    private void onRequestFailed(
+            Request<?> request,
+            OnRequestComplete callback,
+            IOException exception,
+            long requestStartMs,
+            @Nullable HttpResponse httpResponse,
+            @Nullable byte[] responseContents) {
+        try {
+            NetworkUtility.handleException(
+                    request,
+                    callback,
+                    exception,
+                    requestStartMs,
+                    httpResponse,
+                    responseContents,
+                    this);
+        } catch (VolleyError volleyError) {
+            callback.onError(volleyError);
+        }
```

Comment:
You need a callback in the non-VolleyError case

Reply:
Now that the callback is called outside of the handleException method, I think it's clearer to the reader that it's always being thrown.